File: LeetCode/[M] - 2130. Maximum Twin Sum of a Linked List/solution.go
package main

// ListNode - Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func pairSum(head *ListNode) int {
	prev := ListNode{Next: head}
	slow, fast := &prev, &prev
	twins := []int{}
	for fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		twins = append(twins, slow.Val)
	}
	slow = slow.Next
	twin := len(twins) - 1
	maxSum := 0
	for slow != nil {
		currSum := slow.Val + twins[twin]
		if currSum > maxSum {
			maxSum = currSum
		}
		twin--
		slow = slow.Next
	}
	return maxSum
}

File: LeetCode/[M] - 2130. Maximum Twin Sum of a Linked List/solution.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        prev = ListNode(next=head)
        slow, fast = prev, prev
        half = []
        while fast.next:
            slow = slow.next
            fast = fast.next.next
            half.append(slow.val)
        twin = len(half) - 1
        slow = slow.next
        maxSum = 0
        while slow:
            maxSum = max(maxSum, slow.val + half[twin])
            twin -= 1
            slow = slow.next
        return maxSum
